### PR TITLE
Wavelength batching within `amplitudes_for_fields`

### PR DESCRIPTION
**Pull request by smartalecH**
Currently, the `amplitudes_for_fields` function doesn't support wavelength batching. We add that functionality, along with a few additional tests for both single wavelength and multiwavelength cases.

### CODE REVIEW

**Comment by smartalecH** [comment]
(cc @mfschubert )

File: examples/crystal.py [hunk 1]
```diff
@@ -571,10 +611,26 @@ def plot_gaussian_fields(
...
 
     plt.subplots_adjust(left=0, bottom=0, right=1, top=1)
-
     plt.savefig("crystal_gaussian.png", bbox_inches="tight")
 
 
 if __name__ == "__main__":
-    plot_dipole_fields()
-    plot_gaussian_fields()
+    # plot_dipole_fields()
```
Comment (mfschubert) [inline examples/crystal.py:618]: I presume this was for testing --- do you want to restore the calls to the plotting functions?

File: examples/crystal.py [hunk 2]
```diff
@@ -269,6 +271,9 @@ def simulate_crystal_with_gaussian_beam(
         (hx, hy, hz), (x, y, z))`. The fields are returned for an xz slice centered
         on the incident beam.
     """
+    wavelengths = jnp.expand_dims(jnp.atleast_1d(wavelengths), axis=(1, 2))
```
Comment (mfschubert) [inline examples/crystal.py:274]: We actually have a function in the utils module, `atleast_nd` which may be simpler to use here.

File: examples/crystal.py [hunk 3]
```diff
@@ -289,28 +294,42 @@ def simulate_crystal_with_gaussian_beam(
...
         fmm.eigensolve_isotropic_media,
-        wavelength=jnp.asarray(wavelength),
+        wavelength=jnp.asarray(wavelengths),
         in_plane_wavevector=in_plane_wavevector,
         primitive_lattice_vectors=primitive_lattice_vectors,
         expansion=expansion,
         formulation=fmm.Formulation.FFT,
     )
 
+    def _reshape_permittivity(arr: jnp.ndarray):
```
Comment (mfschubert) [inline examples/crystal.py:318]: Is this needed? The automatic broadcasting might already take care of this.

File: examples/crystal.py [hunk 4]
```diff
@@ -327,11 +346,28 @@ def simulate_crystal_with_gaussian_beam(
     # beam, and then solving for the eigenmodes.
     # TODO: replace paraxial Gaussian with a more rigorous expression.
 
-    def _paraxial_gaussian_field_fn(x, y, z):
+    def _paraxial_gaussian_field_fn(
+        x,
+        y,
+        z,
```
Comment (mfschubert) [inline examples/crystal.py:352]: Drop trailing comma to clean this up

File: examples/crystal.py [hunk 5]
```diff
@@ -327,11 +346,28 @@ def simulate_crystal_with_gaussian_beam(
...
+    ):
         # Returns the fields of a z-propagating, x-polarized Gaussian beam.
         # See https://en.wikipedia.org/wiki/Gaussian_beam
-        k = 2 * jnp.pi / wavelength
-        z_r = jnp.pi * beam_waist**2 * jnp.sqrt(permittivity_ambient) / wavelength
+
+        # Adjust array dimensions for proper batching
+        batch_axes = (WAVELENGTH_AXIS, WAVELENGTH_AXIS + 1, WAVELENGTH_AXIS + 2)
+        x = jnp.expand_dims(x, axis=batch_axes)
```
Comment (mfschubert) [inline examples/crystal.py:359]: I didn't go through this carefully, but can you confirm that you actually need to expand the dims of `(x, y, z)` here? It seems like these will be shape `(nx, ny)`, and so by adding the extra dims to `wavelengths_padded` we already achieve compatibility.

**Review by mfschubert — COMMENTED** [review]
I took a quick pass and made some comments. I think that there may be opportunities to simplify things by relying on automatic broadcasting, but I haven't gone through in detail to confirm this. Happy to chat if there are questions.

File: examples/crystal.py (hunk 2 above)
Comment (smartalecH) [reply examples/crystal.py:274]: Great! Can you choose we side to pad? I know numpy's `atleast_nd` prepends the array, when in this case we'd prefer to append.

File: examples/crystal.py (hunk 4 above)
Comment (smartalecH) [reply examples/crystal.py]: Fixed, thanks

File: examples/crystal.py (hunk 3 above)
Comment (smartalecH) [reply examples/crystal.py]: Fixed, thanks.

File: examples/crystal.py (hunk 1 above)
Comment (smartalecH) [reply examples/crystal.py]: Fixed, thanks

File: fmmax/sources.py [hunk 6]
```diff
@@ -38,22 +38,28 @@ def amplitudes_for_fields(
     Returns:
         The forward and backward wave amplitudes.
     """
-    if not (ex.shape == ey.shape == hx.shape == hy.shape) or ex.ndim != 3:
+
+    batch_shape = layer_solve_result.batch_shape
+    expected_fields_dims = (
```
Comment (mfschubert) [inline fmmax/sources.py:43]: Maybe give this comment its own line, so that this formatting is a bit less awkward?

**Review by mfschubert — APPROVED** [review]
LGTM, just one formatting nit.

**Pull request merged**